feat#3 - 사용자 API 구현

shinmin9812 opened the pull request:
### 변경사항
**AS-IS**
- 회원가입
- 로그인
- 설정 업데이트
- 조회

**TO-BE**
- 테스트 코드 구현

### 테스트
- [x] 테스트 코드
- [x] API 테스트

shinmin9812 committed:
feat#3 - 프로젝트 설정 수정
- JWT 사용을 위한 build.gradle, application.yml 수정

shinmin9812 committed:
feat#3 - Jwt 관리 파일 구현
- WebConfig, SecurityConfig Config 파일 생성
- 인증을 위한 AuthenticationPrincipal 어노테이션 생성
- AuthenticationPrincipalArgumentResolver 구현
- Token 생성하는 AuthTokenCreator, 추출하는 AuthorizationExtractor 구현
- Token 생성 및 검증하는 JwtTokenProvider 구현

shinmin9812 committed:
feat#3 - 사용자 API 관련 DTO 구현
- Token 저장하는 LoginMember
- 회원 가입 시 사용되는 MemberJoinReqDto, 로그인 시 사용되는 MemberLoginReqDto, 사용자 설정 업데이트 시 사용되는 MemberUpdateReqDto 구현
- 사용자 정보를 받는 MemberInfoResDto, 로그인 시 토큰을 받는 MemberTokenResDto 구현

shinmin9812 committed:
feat#3 - ErrorCode 추가
- MEMBER_NOT_FOUND 추가

shinmin9812 committed:
feat#3 - 사용자 API 구현
- MemberRepository, MemberService 인터페이스 구현
- MemberServiceImpl 회원가입, 로그인, 설정 업데이트, 조회 기능 구현, Account가 중복됐는지 확인하기 위한 validateUniqueAccount 구현
- MemberController 회원가입, 로그인, 설정 업데이트, 조회 API 구현

nayoonk928 commented on src/main/java/com/wanted/teamV/exception/ErrorCode.java:
```diff
@@ -12,7 +12,8 @@ public enum ErrorCode {
     EMPTY_AUTHORIZATION_HEADER(HttpStatus.BAD_REQUEST, "인증헤더가 비어있습니다."),
-    FAIL_READ_FILE(HttpStatus.INTERNAL_SERVER_ERROR, "파일 읽기 작업이 실패했습니다.")
+    FAIL_READ_FILE(HttpStatus.INTERNAL_SERVER_ERROR, "파일 읽기 작업이 실패했습니다."),
+    MEMBER_NOT_FOUND(HttpStatus.BAD_REQUEST, "사용자를 찾지 못했습니다."),
```
사용자를 찾지 못하는 거니까 NOT_FOUND가 더 괜찮은 것 같습니다.

nayoonk928 commented on src/main/java/com/wanted/teamV/controller/MemberController.java:
```diff
@@ -1,9 +1,76 @@
+        }
+
+        MemberInfoResDto response = memberService.updateMember(loginMember.id(), lat, lon, recommend);
+        return ResponseEntity.ok(response);
```
위 메서드에서 만약, 처음 lat, lon값을 입력하는 상황이라면 member.getLat() 또는 member.getLon()이 null 이기 때문에 
처음 입력받는 경우, memberUpdateReqDto.lat()/lon() 둘 중 하나라도 null이면 둘 다 입력해달라는 식의 에러처리가 필요할 것 같아요

shinmin9812 replied:
처음 회원가입을 할 때 lat, lon은 null이 아니라 0이 들어가서 처음 입력받는 값은 0이 들어갑니다. 그래서 따로 에러처리는 하지 않았는데 하는 게 좋을까요?

nayoonk928 replied:
초기 값이 0이라면 괜찮을 것 같습니다! 
더 생각났는데 추가로 lat과 lon의 범위도 지정하면 좋을 것 같아요

nswon reviewed (commented):
application-test.yml에서도

```
security:
  jwt:
    token:
      secret-key: dGVhbXZfand0X3NlY3JldF9rZXlfdGVzdA==
      access:
        expire-length: 2592000000 #30일
```

추가해주세요! 그럼 테스트코드가 정상적으로 작동할 것 같아요

shinmin9812 committed:
fix #3 - 설정 업데이트 API 수정
- 설정 업데이트 로직 수정

shinmin9812 committed:
fix #3 - 설정 업데이트 API 수정
- ErrorCode에서 MEMBER_NOT_FOUND 삭제
- MemberServiceImpl MEMBER_NOT_FOUND를 NOT_FOUND로 교체
- MemberController에서 사용하지 않는 import 삭제

shinmin9812 committed:
test #3 - application-test.yml 파일 수정

shinmin9812 committed:
test #3 - 사용자 회원가입 API 테스트 코드 구현

nayoonk928 commented on src/main/java/com/wanted/teamV/service/impl/MemberServiceImpl.java:
```diff
@@ -60,7 +60,7 @@ public Long extractUserId(String accessToken) {
     public MemberInfoResDto getMember(Long memberId) {
-        Member member = memberRepository.findById(memberId).orElseThrow(() -> new CustomException(ErrorCode.MEMBER_NOT_FOUND));
+        Member member = memberRepository.findById(memberId).orElseThrow(() -> new CustomException(ErrorCode.NOT_FOUND));
         MemberInfoResDto response = MemberInfoResDto.builder()
```
음식점이 없는 경우, 멤버가 없는 경우가 있으니 MEMBER_NOT_FOUND를 유지하고 status만 변경하는게 좋을 것 같아요. 제가 위에 헷갈리게 적어놨네요 😢

shinmin9812 committed:
fix #3 - ErrorCode 수정
- MEMBER_NOT_FOUND 추가 및 적용

shinmin9812 committed:
fix #3 - 사용자 회원가입, 설정 업데이트 API 수정
- 회원가입 시 위도, 경도 기본값 null로 설정
- 설정 업데이트 시 위도, 경도값이 비어있을 시 NULL_LAT(LON)_VALUE 에러코드 추가
- 위도, 경도가 범위를 넘어갈 시 INVALID_LAT_RANGE 에러코드 추가

shinmin9812 committed:
test #3 - 사용자 로그인 API 테스트 코드 구현
- Member 엔티티에 테스트용 객체 생성 메소드 구현
- 사용자 로그인 성공, 실패 테스트 코드 구현

shinmin9812 committed:
test #3 - 사용자 조회 API 테스트 코드 구현

shinmin9812 committed:
test #3 - 사용자 설정 업데이트 API 테스트 코드 구현

shinmin9812 committed:
Merge branch 'develop' into feat/#3
